### PR TITLE
{?}[GCCcore/9.3.0] HEASoft v6.29 + v6.30.1

### PR DESCRIPTION
**Pull request by volodymyrss**
(created using `eb --new-pr`)

### CODE REVIEW

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:10]: Could you please make the order of these parts match other easyconfigs, easyblock first, dependencies below sources.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:13]: We really like to provide a homepage and definitely a short description of the software. Is there really nothing available here?

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:15]: You are specifying a very old compiler here, but the dependencies you specify (Perl 5.30.2) only exists for much newer GCCcore.  Unless you have backported a whole lot of easyconfigs to this really old version, it will not work as is (and it will definitely not work in this repository. 
Version 9.3.0 should work fine though (matches with your CFITSIO and Perl dependencies).

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:24]: You'll need to add a build dependency  on binutils 2.34 (matching GCCcore  9.3.0)
```
builddependencies = [('binutils', '2.34')]
```

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:73]: This seems like it would be solved by having just modextrapaths extend `PFILES` in this easyconfig, which will append to the path variable instead of replacing it (which is what modextravars does).
This method with a footer will not work well, because unloading the module will not restore it (and a bunch of other problems like this)

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:77]: I would prefer not to add additional messages unless it's necessary information for how to use a particular module. It would be very annoying if many modules started doing it, so even then, it's done *very* selectively.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:88]: Hm, I don't think we should distribute these instructions? They would assume you have loading other modules prior to this?
Is not the setup done by the module loading sufficient? what else needs to be set up (and why can't we provide this as part of the module)?

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:33]: So, this installpath thing is definfiely the big issue here. First, I think we should try to tackle this at the install step instead, getting rid of this extra directory. This will just make the rest of the easyconfig muuuuch cleaner. No need to specify PATH and LD_LIBRARY_PATH since they will just work. All the modextravars will be much simpler.

Dealing with the extra path can possibly be done with help of a preinstall/postinstall command, or maybe with a small patch to the makefiles that adds this path (perfectly fine to have a EB-specific patch to do things like this if the configure system of this software doesn't allow for this choice.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:54]: You can use `$EBROOTPERL/bin/perl` if you really need a full path. Otherwise just `perl` might suffice.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (Micket) [inline easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:38]: It would be good to specify a few key binaries and/or libraries in the file section here. dirs is less important (since it's usually caught by `files`

**Review by Micket — CHANGES_REQUESTED** [review]
I believe you are a new contributor, so a warm welcome to EasyBuild!
Quite a big first contribution here, so I tried to give it a thorough and detailed review. Feel free to ask question if you are unsure about what I mean here.

**Comment by volodymyrss** [comment]
> I believe you are a new contributor, so a warm welcome to EasyBuild!
> Quite a big first contribution here, so I tried to give it a thorough and detailed review. Feel free to ask question if you are unsure about what I mean here.

Thanks a lot for the kind and welcoming message! 

I am trying to see if I can adapt this meaningfully to EasyBuild, so your suggestions are very welcome. 
Much of the comments were meant primarily for the reviewers.
I will try address and comment on specific points, and re-request review when I am done. Would that be a good workflow?

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: done!

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: Description and homepage added.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:11]: I did back-port about easyconfigs. I had some trouble building, this software can be delicate.
Also I was trying to extrapolate from whatever where was already available on our cluster. 
I will try with 9.3.0, it will take some time.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:91]: We would not want to distribute this. This is a variation used by one our project with lmod before. I wanted to ask you how to deal with this situation.
The problem is that the initialization which is guaranteed to be working in different situations by the package developers, and will probably remains so for a while is this one:
```bash
$ . $HEADAS/headas-init.sh
```
Anything else: particular set of variables set,  is just me knowing what it does to some degree.
However, in some cases it will set different variables.
Should I implement the entire logic hidden in  `$HEADAS/headas-init.sh` in the easyconfig? Is it ok to make an assumption?

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb:80]: sure, makes sense

**Comment by Micket** [comment]
Go for it, let me know when you are ready for round 2!

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: Updated, thanks.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: Removed.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: moved to 9.3.0, works fine. Thanks for the suggestion!

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: Ok I delete this.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: Is it possible to get the user home path?

Here I really need to replace `PFILES`, not append. So I would put it to `modextravars`.
But maybe there is a way to use user home in the module?

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: Ok, just `perl` works. Thanks!

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: sure, added some `files`: [there](https://github.com/volodymyrss/easybuild-easyconfigs/blob/c23511f9a6e4be4390db483d8fa185d1af4e2794/easybuild/easyconfigs/h/HEASoft/HEASoft-6.29-GCCcore-9.3.0.eb#L52-L55)

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: > This method with a footer will not work well, because unloading the module will not restore it (and a bunch of other problems like this)

Concerning unloading, I thought `pushenv` (unlike `setenv`) will preserve previous variable state and restore it on unload, according to the [doc](https://lmod.readthedocs.io/en/latest/050_lua_modulefiles.html):

```
pushenv (“NAME”, “value”):
    sets NAME to value just like setenv. In addition it saves the previous value in a hidden environment variable. This way the previous state can be returned when a module is unloaded. pushenv (“FOO”,false) will clear “FOO” and the pop will return the previous value.
```
(there is also a [tcl version](https://lmod.readthedocs.io/en/latest/051_tcl_modulefiles.html))

I tried it and it seems to work: the variable is restored to its pre-load value with unload. 

But I am not really experienced with modules and easybuild. I also note that no other easyconfig seems to use pushenv.
So I would be happy to use a better, more accepted, solution if there is one. 
I just do not know how else to set variable including dynamic `$HOME` path.

File: easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb
Comment (volodymyrss) [reply easybuild/easyconfigs/h/HEASoft/HEASoft-6.27.2-GCCcore-4.9.3.eb]: Ok, I made a patch. Much cleaner indeed!

**Comment by volodymyrss** [comment]
> Go for it, let me know when you are ready for round 2!

I think I am ready! 

There is one remaining point: if it's possible to use `pushenv`, and if not - how to set variable to path in `HOME`. It still seems to me that `pushenv` solution satisfies modules behavior. But I may be wrong, and I welcome other suggestions.

I left couple of comments on issues inside which may be useful for the users.

Also, I added another version, with `foss`. I did not know how else to add a `numpy` dependency. This enables python integration, which is critical for some people, and completely irrelevant for other.
But `foss` brings a lot more dependencies, with some peculiar issues in some versions.
So I was wondering if it's possible to parametrize it somehow instead of making two very similar eb's.

**Comment by Micket** [comment]
F.Y.I. there are still things that must be addressed before we can merge this, but I must admit I simply haven't had time (and won't for the forseeable future) due to personnel shortage. I've mentioned this to the other maintainers so maybe someone else will have a look at this.

**Comment by boegel** [comment]
closing this since `GCC/9.3.0` is deprecated, see https://docs.easybuild.io/policies/toolchains

@volodymyrss I'm sorry for not coming back to this...

If `HEASoft` is still relevant, please consider opening a new pull request with a more recent toolchain

**Comment by volodymyrss** [comment]
> closing this since `GCC/9.3.0` is deprecated, see https://docs.easybuild.io/policies/toolchains
> 
> @volodymyrss I'm sorry for not coming back to this...
> 
> If `HEASoft` is still relevant, please consider opening a new pull request with a more recent toolchain

It's relevant, it's used in new missions telescopes, especially NASA. 
But, I am totally sure if it's when I invest a bit of effort in new PR it will be reviewed. What do you think?

**Pull request closed without merging**
